pkg/volume/rbd/rbd_util.go

```diff
@@ -256,30 +357,40 @@ func (util *RBDUtil) AttachDisk(b rbdMounter) (string, error) {
 			}
 			return !used, nil
 		})
-		// return error if rbd image has not become available for the specified timeout
+		// Return error if rbd image has not become available for the specified timeout.
 		if err == wait.ErrWaitTimeout {
 			return "", fmt.Errorf("rbd image %s/%s is still being used", b.Pool, b.Image)
 		}
-		// return error if any other errors were encountered during wating for the image to becme avialble
+		// Return error if any other errors were encountered during wating for the image to become available.
 		if err != nil {
 			return "", err
 		}
 
 		mon := util.kernelRBDMonitorsOpt(b.Mon)
 		glog.V(1).Infof("rbd: map mon %s", mon)
-		if b.Secret != "" {
-			output, err = b.exec.Run("rbd",
-				"map", b.Image, "--pool", b.Pool, "--id", b.Id, "-m", mon, "--key="+b.Secret)
-		} else {
-			output, err = b.exec.Run("rbd",
-				"map", b.Image, "--pool", b.Pool, "--id", b.Id, "-m", mon, "-k", b.Keyring)
+
+		if nbdToolsFound {
+			output, err = execRbdMap(b, "rbd-nbd", mon)
+			if err != nil {
+				glog.Infof("rbd-nbd: map error %v, rbd-nbd output: %s", err, string(output))
+				glog.V(4).Info("will retry using rbd after rbd-nbd failure")
+			} else {
+				devicePath, mapped = waitForPath(b.Pool, b.Image, 10 /*maxRetries*/, true /*useNbdDriver*/)
+			}
 		}
-		if err != nil {
-			glog.V(1).Infof("rbd: map error %v, rbd output: %s", err, string(output))
-			return "", fmt.Errorf("rbd: map failed %v, rbd output: %s", err, string(output))
+		if !mapped {
+			_, err = b.exec.Run("modprobe", "rbd")
+			if err != nil {
+				glog.Warningf("rbd: failed to load rbd kernel module: %v", err)
+			}
+			output, err = execRbdMap(b, "rbd", mon)
+			if err != nil {
+				glog.V(1).Infof("rbd: map error %v, rbd output: %s", err, string(output))
+				return "", fmt.Errorf("rbd: map failed %v, rbd output: %s", err, string(output))
```

Comment:
please also return rbd-nbd output

Reply:
Would you recommend that we merge the error messages (including output from both commands) for the case when both fail?

Reply:
yes, please do - picking up the silent errors helps diagnosis.